File: src/engine/deck.py
"""
Deck extraction and requirement-building.

Extracted from deck_cost.py. The raw-board extraction logic
(extract_raw_board_cards / extract_deck_records) is behaviorally
unchanged from the original implementation.

build_requirements() gained one addition beyond the original: an
`include_commander` flag and a `board` field on each CardRequirement.
Neither changes the numbers produced when include_commander=True (the
default, and the only behavior deck_cost.py ever had) -- see the
include_commander note in engine.py for why this isn't wired to
Moxfield's own `includeCommandersInPrice` flag yet.
"""

from __future__ import annotations

from typing import Any

from .models import CardRequirement


def qty(record: dict[str, Any]) -> int:
    try:
        return max(int(record.get("quantity", 1)), 0)
    except (TypeError, ValueError):
        return 1


# ---------------------------------------------------------------------
# NORMALIZED DECK
# ---------------------------------------------------------------------

def normalized_deck_records(deck: dict[str, Any]) -> list[dict[str, Any]]:
    """Return commander + mainboard records; deliberately excludes sideboard."""
    records = []
    for section in ("commander", "mainboard"):
        value = deck.get(section, [])
        if isinstance(value, list):
            records.extend(x for x in value if isinstance(x, dict))
    return records


def build_requirements(
    normalized_deck: dict[str, Any],
    include_commander: bool = True,
) -> list[CardRequirement]:
    """
    Build the list of CardRequirement objects the matcher needs.

    include_commander=True (the default) reproduces deck_cost.py's
    original, unconditional behavior of counting the commander as
    required. Setting it False excludes commander records entirely.
    """
    sections = ("commander", "mainboard") if include_commander else ("mainboard",)

    requirements: list[CardRequirement] = []
    for section in sections:
        value = normalized_deck.get(section, [])
        if not isinstance(value, list):
            continue
        for record in value:
            if not isinstance(record, dict) or not record.get("name"):
                continue
            requirements.append(
                CardRequirement(
                    name=str(record["name"]),
                    quantity=qty(record),
                    board=section,
                    finish=record.get("finish", "nonFoil"),
                    scryfall_id=record.get("scryfall_id"),
                )
            )
    return requirements


# ---------------------------------------------------------------------
# RAW MOXFIELD DECK
# ---------------------------------------------------------------------

def extract_raw_board_cards(board: Any) -> list[dict[str, Any]]:
    """
    Raw Moxfield board shape:

        {
          "count": 99,
          "cards": {
             "internal-id": {
                "quantity": 1,
                "boardType": "mainboard",
                "finish": "nonFoil",
                "card": {...}
             }
          }
        }

    Flatten each nested card into a convenient record while preserving
    the board-specific quantity/finish fields.
    """
    if not isinstance(board, dict):
        return []

    cards = board.get("cards", {})
    if isinstance(cards, dict):
        entries = cards.values()
    elif isinstance(cards, list):
        entries = cards
    else:
        return []

    result = []

    for entry in entries:
        if not isinstance(entry, dict):
            continue

        card = entry.get("card")
        if not isinstance(card, dict):
            continue

        record = dict(card)
        record["quantity"] = qty(entry)
        record["finish"] = entry.get(
            "finish",
            card.get("defaultFinish", "nonFoil"),
        )
        record["boardType"] = entry.get("boardType")
        record["isFoil"] = bool(entry.get("isFoil", False))
        record["isAlter"] = bool(entry.get("isAlter", False))
        record["isProxy"] = bool(entry.get("isProxy", False))
        result.append(record)

    return result


def extract_deck_records(raw_deck: dict[str, Any]) -> list[dict[str, Any]]:
    """
    Extract the actual Commander deck from a raw Moxfield response.

    IMPORTANT:
      raw Moxfield does NOT use top-level "commander"/"mainboard".
      It uses:

          raw_deck["boards"]["commanders"]["cards"]
          raw_deck["boards"]["mainboard"]["cards"]

      The raw response also has a sideboard. It is intentionally excluded.
    """
    if not isinstance(raw_deck, dict):
        raise TypeError(
            f"Unable to interpret raw deck data: {type(raw_deck).__name__}"
        )

    boards = raw_deck.get("boards")
    if not isinstance(boards, dict):
        raise TypeError(
            "Raw Moxfield deck is missing the expected 'boards' object."
        )

    commander = extract_raw_board_cards(boards.get("commanders"))
    mainboard = extract_raw_board_cards(boards.get("mainboard"))

    records = commander + mainboard

    if not records:
        raise TypeError(
            "Raw Moxfield deck contained boards, but no commander/mainboard "
            "cards could be extracted."
        )

    return records

File: src/engine_js/deck.js
/**
 * deck.js
 * -------
 * Port of src/engine/deck.py -- no behavior change.
 *
 * extractRawBoardCards / extractDeckRecords are direct translations of
 * the Python originals. buildRequirements carries the same
 * includeCommander flag and board field as the Python version -- see
 * engine.js for the note on why that flag doesn't read Moxfield's own
 * includeCommandersInPrice field yet.
 */

const { qty } = require("./util");

// ---------------------------------------------------------------------
// NORMALIZED DECK
// ---------------------------------------------------------------------

/**
 * Return commander + mainboard records; deliberately excludes sideboard.
 * @param {object} deck
 * @returns {Array<object>}
 */
function normalizedDeckRecords(deck) {
  const records = [];
  for (const section of ["commander", "mainboard"]) {
    const value = deck[section];
    if (Array.isArray(value)) {
      for (const x of value) {
        if (x !== null && typeof x === "object") records.push(x);
      }
    }
  }
  return records;
}

/**
 * Build the list of CardRequirement objects the matcher needs.
 *
 * includeCommander=true (the default) reproduces deck_cost.py's
 * original, unconditional behavior of counting the commander as
 * required. Setting it false excludes commander records entirely.
 *
 * @param {object} normalizedDeck
 * @param {boolean} [includeCommander=true]
 * @returns {import('./models').CardRequirement[]}
 */
function buildRequirements(normalizedDeck, includeCommander = true) {
  const sections = includeCommander ? ["commander", "mainboard"] : ["mainboard"];

  const requirements = [];
  for (const section of sections) {
    const value = normalizedDeck[section];
    if (!Array.isArray(value)) continue;
    for (const record of value) {
      if (record === null || typeof record !== "object" || !record.name) continue;
      requirements.push({
        name: String(record.name),
        quantity: qty(record),
        board: section,
        finish: record.finish || "nonFoil",
        scryfallId: record.scryfall_id ?? null,
      });
    }
  }
  return requirements;
}

// ---------------------------------------------------------------------
// RAW MOXFIELD DECK
// ---------------------------------------------------------------------

/**
 * Raw Moxfield board shape:
 *
 *   {
 *     count: 99,
 *     cards: {
 *       "internal-id": {
 *         quantity: 1,
 *         boardType: "mainboard",
 *         finish: "nonFoil",
 *         card: {...}
 *       }
 *     }
 *   }
 *
 * Flatten each nested card into a convenient record while preserving
 * the board-specific quantity/finish fields.
 * @param {object} board
 * @returns {Array<object>}
 */
function extractRawBoardCards(board) {
  if (board === null || typeof board !== "object") return [];

  const cards = board.cards;
  let entries;
  if (cards !== null && typeof cards === "object" && !Array.isArray(cards)) {
    entries = Object.values(cards);
  } else if (Array.isArray(cards)) {
    entries = cards;
  } else {
    return [];
  }

  const result = [];

  for (const entry of entries) {
    if (entry === null || typeof entry !== "object") continue;

    const card = entry.card;
    if (card === null || typeof card !== "object") continue;

    const record = { ...card };
    record.quantity = qty(entry);
    record.finish = entry.finish ?? card.defaultFinish ?? "nonFoil";
    record.boardType = entry.boardType;
    record.isFoil = Boolean(entry.isFoil);
    record.isAlter = Boolean(entry.isAlter);
    record.isProxy = Boolean(entry.isProxy);
    result.push(record);
  }

  return result;
}

/**
 * Extract the actual Commander deck from a raw Moxfield response.
 *
 * IMPORTANT:
 *   raw Moxfield does NOT use top-level "commander"/"mainboard".
 *   It uses:
 *
 *       rawDeck.boards.commanders.cards
 *       rawDeck.boards.mainboard.cards
 *
 *   The raw response also has a sideboard. It is intentionally excluded.
 *
 * @param {object} rawDeck
 * @returns {Array<object>}
 */
function extractDeckRecords(rawDeck) {
  if (rawDeck === null || typeof rawDeck !== "object") {
    throw new TypeError(`Unable to interpret raw deck data: ${typeof rawDeck}`);
  }

  const boards = rawDeck.boards;
  if (boards === null || typeof boards !== "object") {
    throw new TypeError("Raw Moxfield deck is missing the expected 'boards' object.");
  }

  const commander = extractRawBoardCards(boards.commanders);
  const mainboard = extractRawBoardCards(boards.mainboard);

  const records = [...commander, ...mainboard];

  if (records.length === 0) {
    throw new TypeError(
      "Raw Moxfield deck contained boards, but no commander/mainboard cards could be extracted."
    );
  }

  return records;
}

module.exports = {
  normalizedDeckRecords,
  buildRequirements,
  extractRawBoardCards,
  extractDeckRecords,
};
